src/RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using RoslynAnalyzer;
+using System;
+using System.Collections.Immutable;
+
+namespace ILTrimmingAnalyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class RequiresUnreferencedCodeAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "IL2026";
+
+        private const string s_trimmingEnabledString = "PublishTrimmed";
+
+        private static readonly LocalizableString s_title = new LocalizableResourceString(
+            nameof(RequiresUnreferencedCodeAnalyzer) + "Title",
+            Resources.ResourceManager,
+            typeof(Resources));
+        private static readonly LocalizableString s_messageFormat = new LocalizableResourceString(
+            nameof(RequiresUnreferencedCodeAnalyzer) + "Message",
+            Resources.ResourceManager,
+            typeof(Resources));
+        private const string s_category = "Trimming";
+
+        private static DiagnosticDescriptor s_rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            s_title,
+            s_messageFormat,
+            s_category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var isTrimmingEnabled = context.Options.GetMSBuildPropertyValue(
+                    s_trimmingEnabledString, compilation, context.CancellationToken);
+
+                if (!string.Equals(isTrimmingEnabled?.Trim(), "true", StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(operationContext =>
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    var method = invocation.TargetMethod;
+                    var attributes = method.GetAttributes();
+
+                    foreach (var attr in attributes)
+                    {
+                        if (attr.AttributeClass is { } attrClass &&
```

Comment:
Perf nit: this will full scan all attributes on all called methods - for a reasonably sized assembly this will perform a lot of comparisons. I wonder if there's some way to "cache" some information to prevent such full scans.
Linker has even bigger potential problem (as it processes lot more code at once) - it solves it by storing per method cache of "linker attributes" (basically attributes recognized by linker). Since almost all methods have this cache empty, it's very fast to check that there's nothing to do.

I don't know if there's some typical pattern how roslyn analyzers solve this problem (or if they actually solve it at all).

Reply:
A cache here would probably be fine, but symbols are not guaranteed reference equality in Roslyn, so in cases where there is no reference-equal member in the cache, it may be faster to loop over the attributes instead of comparing the symbols.